mqtt-streaming/src/main/scala/akka/stream/alpakka/mqtt/streaming/impl/ClientConnector.scala

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (C) 2016-2018 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.mqtt.streaming
+package impl
+
+import akka.actor.typed.{ActorRef, Behavior}
+import akka.actor.typed.scaladsl.Behaviors
+import akka.annotation.InternalApi
+import akka.util.ByteString
+
+/*
+ * A client connector is a Finite State Machine that manages MQTT client
+ * session state. A client connects to a server, subscribes/unsubscribes
+ * from topics to receive publications on and publishes to its own topics.
+ */
+@InternalApi private[streaming] object ClientConnector {
+
+  type ConnectData = Option[_]
+
+  /*
+   * Construct with the starting state
+   */
+  def apply(settings: MqttSessionSettings): Behavior[Event] =
+    disconnected(Uninitialized(settings))
+
+  // Our FSM data and events
+
+  sealed abstract class Data(val settings: MqttSessionSettings)
+  final case class Uninitialized(override val settings: MqttSessionSettings) extends Data(settings)
+  final case class ConnectReceived(connect: Connect,
+                                   connectData: ConnectData,
+                                   override val settings: MqttSessionSettings)
+      extends Data(settings)
+  final case class ConnAckReceived(connect: Connect, connAck: ConnAck, override val settings: MqttSessionSettings)
+      extends Data(settings)
+
+  sealed abstract class Event
+  final case class ConnectReceivedLocally(connect: Connect, connectData: ConnectData, remote: ActorRef[ByteString])
+      extends Event
+  final case class ConnAckReceivedFromRemote(connAck: ConnAck, local: ActorRef[(ConnAck, ConnectData)]) extends Event
+  case object ReceiveConnAckTimeout extends Event
+  case object LostConnection extends Event
+  case object DisconnectReceivedLocally extends Event
+
+  // State event handling
+
+  def disconnected(data: Uninitialized): Behavior[Event] = Behaviors.receiveMessagePartial {
+    case ConnectReceivedLocally(connect, connectData, remote) =>
+      forwardConnectToRemote(connect, remote)
+      serverConnect(ConnectReceived(connect, connectData, data.settings))
+  }
+
+  def serverConnect(data: ConnectReceived): Behavior[Event] = Behaviors.withTimers { timer =>
```

Comment:
I don't know the MQTT domain well, so this might be nonsense, but for a state something like "connecting" or "waitingForConnAck" might be a better name.

Reply:
I think the current name makes sense. From my state diagram:

![image](https://user-images.githubusercontent.com/694893/46263910-2945c680-c559-11e8-99db-bb6065724b66.png)